### ISSUE TITLE
FSE: Most site building blocks need previews and placeholders

### ISSUE THREAD

**Issue by jameskoster** [WordPress/gutenberg#30029]
- [x] Site Logo - #35588
- [x] Site Tagline
- [x] Site Title - #35590
- [x] Login/out - https://github.com/WordPress/gutenberg/pull/62937
- [x] Post Title - https://github.com/WordPress/gutenberg/pull/62955
- [x] Post Content - https://github.com/WordPress/gutenberg/pull/62968
- [x] Post Author - https://github.com/WordPress/gutenberg/pull/62978
- [ ] Post Comments Count 
- [x] Post Date - https://github.com/WordPress/gutenberg/pull/63368
- [x] Post Excerpt - https://github.com/WordPress/gutenberg/pull/63010
- [x] Post Featured Image - https://github.com/WordPress/gutenberg/pull/63011
- [x] Post Tags - https://github.com/WordPress/gutenberg/pull/63369
- [ ] Term Description 
- [x] Posts List - https://github.com/WordPress/gutenberg/pull/63286
- [ ] Archive Title 
- [ ] Post Categories 
- [ ] Next post 
- [ ] Previous post 

[Here's a figma file](https://www.figma.com/file/xMD1TfdiAl6nmS9opIlUya/Site-building-block-placeholders-previews?node-id=1%3A18) where we can work on this.

**Comment by jameskoster** [comment]
Some initial ideas for simplest text-based blocks:

| Block | Placeholder |
| --- | --- |
| Site Tagline | The site description  |
| Site Title | The site title  |
| Login/out | Log in |
| Post Title | The post title |
| Post Content | The content of a post, page, or any other content-type. It may include a variety of different blocks like paragraphs, lists, image galleries, and more. |
| Post Author | The post author  |
| ~Post Comment~ |  ~The content of a comment.~ |
| Post Comments Count | 1 |
| Post Date | [Gutenberg release date?] |
| Post Excerpt | The excerpt of a post, page, or any other content-type. |
| Post Tags | <a href="#">Tag</a> |
| Term Description | The description of a tag, category, or other custom taxonomy. |
| Archive Title | The title of an archive like a category, tag, author, or date |
| Post Categories | <a href="#">Category</a> |
| Next post | <a href="#">Next post</a> |
| Previous post | <a href="#">Previous post</a> |

**Comment by carolinan** [comment]
I am wondering about this `Login/out | Log in` because in the editor we are always logged in, should it not be "Log out"?

**Comment by jameskoster** [comment]
I was thinking about it from the hypothetical visitors perspective, who in most cases will be logged out. It's not a strong feeling though. "Log out" works too :)

**Comment by jameskoster** [comment]
> Post Title | The post title

I wonder what scope there is for these "post" block placeholders to be dynamic. While thinking about the template mosaic view for instance, it would be potentially confusing to see "The post title" in the page template.

If not dynamic, they perhaps need to be even more generic. Consider the singular template which can display posts _and_ pages. What should the Post Title block placeholder display in that case? Maybe something simpler like "The title" would work better?

**Comment by jameskoster** [comment]
Excluded some of the comment related blocks as these will need to be broken down in to smaller pieces (see https://github.com/WordPress/gutenberg/issues/24101).

**Comment by jameskoster** [comment]
Is there any reason not to use the WordPress logo for the Site Logo block preview (when no logo is set of course)?

<img width="688" alt="Screenshot 2021-04-12 at 15 48 54" src="https://user-images.githubusercontent.com/846565/114414359-a597d680-9ba6-11eb-90ac-a4a2818e5e47.png">

**Comment by jameskoster** [comment]
I worked on designs for the text-based site blocks in [figma](https://www.figma.com/file/xMD1TfdiAl6nmS9opIlUya/Site-building-block-placeholders-previews?node-id=0%3A1). I removed the Post Comments Form, Query, and Template Part blocks from the OP for now as they are bit more complex and I suspect they will need dedicated issues to explore in detail.

### Site Logo
<img width="883" alt="Screenshot 2021-04-12 at 16 56 03" src="https://user-images.githubusercontent.com/846565/114424418-feb83800-9baf-11eb-8976-dd7c38c0a40f.png">


---

### Site Title
<img width="1231" alt="Screenshot 2021-04-12 at 16 57 01" src="https://user-images.githubusercontent.com/846565/114424535-20b1ba80-9bb0-11eb-8f1b-4c213a313890.png">


---

### Site Tagline
<img width="1259" alt="Screenshot 2021-04-12 at 16 57 46" src="https://user-images.githubusercontent.com/846565/114424645-3cb55c00-9bb0-11eb-8366-9ff51cd83173.png">


---

### Login/out
<img width="920" alt="Screenshot 2021-04-12 at 16 58 14" src="https://user-images.githubusercontent.com/846565/114425132-b2b9c300-9bb0-11eb-9288-c783581987f0.png">


---

### Next / Previous post
<img width="1241" alt="Screenshot 2021-04-12 at 16 58 32" src="https://user-images.githubusercontent.com/846565/114425150-b77e7700-9bb0-11eb-934e-1feaafb064a3.png">


---

### Post Title
<img width="1248" alt="Screenshot 2021-04-12 at 16 58 45" src="https://user-images.githubusercontent.com/846565/114425173-bbaa9480-9bb0-11eb-910e-309c8fb99006.png">


---

### Post Content
<img width="1253" alt="Screenshot 2021-04-12 at 16 59 02" src="https://user-images.githubusercontent.com/846565/114425208-c1a07580-9bb0-11eb-8fb7-d3f19ef2631f.png">


---

### Post Author
<img width="1261" alt="Screenshot 2021-04-12 at 16 59 19" src="https://user-images.githubusercontent.com/846565/114425234-c82eed00-9bb0-11eb-88f8-3c772f966f85.png">


---

### Post Comments Count
<img width="1273" alt="Screenshot 2021-04-12 at 16 59 34" src="https://user-images.githubusercontent.com/846565/114425257-ce24ce00-9bb0-11eb-9eed-21d7f98becb3.png">


---

### Post Date
<img width="1263" alt="Screenshot 2021-04-12 at 16 59 45" src="https://user-images.githubusercontent.com/846565/114425274-d2e98200-9bb0-11eb-8317-50fc8fdb2bdd.png">


---

### Post Excerpt
<img width="1242" alt="Screenshot 2021-04-12 at 16 59 57" src="https://user-images.githubusercontent.com/846565/114425303-da109000-9bb0-11eb-9176-d9b0ae258a5c.png">


---

### Post Tags
<img width="1245" alt="Screenshot 2021-04-12 at 17 00 09" src="https://user-images.githubusercontent.com/846565/114425322-df6dda80-9bb0-11eb-9a5f-dacbbf80afb6.png">


---

### Post Categories
<img width="1252" alt="Screenshot 2021-04-12 at 17 00 21" src="https://user-images.githubusercontent.com/846565/114425360-e8f74280-9bb0-11eb-9d77-bfc553e4c760.png">


---

### Archive Title
<img width="1270" alt="Screenshot 2021-04-12 at 17 00 34" src="https://user-images.githubusercontent.com/846565/114425387-ef85ba00-9bb0-11eb-9c99-415a0eb7811b.png">


---

### Term Description
<img width="1265" alt="Screenshot 2021-04-12 at 17 00 52" src="https://user-images.githubusercontent.com/846565/114425408-f44a6e00-9bb0-11eb-81d5-96225dece8b8.png">


---

**Comment by carolinan** [comment]
Do we just add an empty example in index.js to these blocks, and adjust the width for some?

I can't figure out how to create an example as shown in the images.
For example, the site tagline shows the placeholder text from edit.js, and I don't think that text should be changed to "Just another WordPress site"

**Comment by jameskoster** [comment]
We've recently updated the placeholders, so we might want to close this issue in favor of one that concentrates on the previews.

The Site Tagline block for example still doesn't have a preview:

<img width="678" alt="Screenshot 2021-10-22 at 13 44 34" src="https://user-images.githubusercontent.com/846565/138456034-a1b942ba-e35c-43f7-b37f-74c7564b6439.png">

**Comment by carolinan** [comment]
The comment I added above is all about that preview. So a new issue wont help me contribute towards making the preview look like in the image :).

Adding a width to block.json, or the empty example to index.js, displays the current tagline.
If there is no tagline it displays the placeholder text from edit.js which is "Write site tagline…".

**Comment by jameskoster** [comment]
Ah sorry, I think I misread your comment.

> Adding a width to block.json, or the empty example to index.js, displays the current tagline. If there is no tagline it displays the placeholder text from edit.js which is "Write site tagline…".

Is this definitely the behaviour we want, contextual previews? Given our current handling of placeholders the previews could end up being a bit lacklustre. Like the "Write site tagline..." example you shared. As another example, if a featured image isn't set you would end up seeing something like this:

<img width="640" alt="Screenshot 2021-10-22 at 14 00 37" src="https://user-images.githubusercontent.com/846565/138457916-ee1e2f57-59d7-4db1-86c3-0bcd25ed5511.png">

An alternative would be to hard code generic previews – which is how previews for regular blocks work. So when I hover Featured Image I would just see an image in the preview, regardless of whether one is set or not.

**Comment by carolinan** [comment]
But how? :)
For example, this is the paragraph block. The post/site blocks do not have the "content" attribute.
```
	example: {
		attributes: {
			content: __(
				'In a village of La Mancha, the name of which I have no desire to call to mind, there lived not long since one of those gentlemen that keep a lance in the lance-rack, an old buckler, a lean hack, and a greyhound for coursing.'
			),
			style: {
				typography: {
					fontSize: 28,
				},
			},
			dropCap: true,
		},
	},
```

**Comment by carolinan** [comment]
And I can't find any documentation for the "example" setting in block creation. Guess what shows up if I search for block.json example 😆

**Comment by jameskoster** [comment]
> But how? :)

Good question :D I'm sorry to say I don't know. Perhaps we need to enhance the `example` functionality in order to account for these data-driven blocks?

**Comment by carolinan** [comment]
Yes we probably need to invent something new.

**Comment by talldan** [comment]
> Yes we probably need to invent something new.

I stumbled across this issue. Unless this has been solved since your comment, I think we do need something new. There are probably quite a few different ways to solve this, but none of them that I could think of are ideal.

I think the third option I listed below would be my preferred solution. Any other options anyone can think of?

#### Option 1. Mock block context and entity data in the block definition
These blocks use block context to indicate which post they should read from. It could be possible to mock this data as part of the block definition:
```js
example: {
    context: {
        postType: 'example',
        postId: '1',
    },
    entities: [
        {
            type: 'example',
            id: '1',
            featuredImage: 'https://wordpress.org/example-image.png'
        }
    ],
    attributes: { // ... },
    innerBlocks: [ // ... ],
}
```

Problems:
- It looks easy, but this could be technically difficult to implement. It'll be difficult to control the context for inner blocks, because the preview is rendered as a single `BlockList`.
- Bad developer experience

#### Option 2. Overriding values in the block `edit`
The simplest option could be to pass example data into the block edit component and use that to override the values of particular variables.

block.json
```json
{
    "example": {
        "src": "https://wordpress.org/example-image.png"
    },
```


```js
function PostFeaturedImageEdit( { attributes, example: { src: exampleSrc }, ... } ) {
    const isBlockPreview = useIsBlockPreview();
    let [ imageSrc ] = useEntityProp( // ... );
    
    if ( isBlockPreview ) {
        imageSrc = exampleSrc;
    }

    // ...
}
```

Problems:
- This is pretty simple, but I don't think it's good practice to introduce extra logic to block rendering just for examples. I think it'll reduce code quality.

#### Option 3. Refactor blocks so that we can create `BlockExample` components
Blocks could have their presentation logic separated from their business logic. It'd then be easier to create a `BlockExample` component that renders the same thing as `BlockEdit` but with dummy data:
edit.js
```js
function PostFeaturedImageEdit( { attributes, ... } ) {
    const [ imageSrc ] = useEntityProp( // ... );
   
    return <PostFeaturedImageEditView src={ imageSrc } />
}
```

example.js
```js
function PostFeaturedImageExample() {
   return <PostFeaturedImageEditView src="https://wordpress.org/example-image.png" />
}
```

index.js
```js
// A component can be declared as part of the example.
example: {
    component: PostFeaturedImageExample
    viewportWidth: 500
}
```

Problems:
- Simple, but refactoring each block could be a lot of work
- Could be open to abuse - e.g. rendering something in the preview that shouldn't be there. 🤷 
- This option would never show 'real' data when it's available.

**Comment by jameskoster** [comment]
Since this issue was opened we've done quite a bit of work around improving the on-canvas representation of these blocks, accounting for context. For example the Featured Image block now displays a placeholder 'graphic' when appropriate. 

So perhaps these block previews should just render the block itself (hopefully with the context passed to it). Is that an option?

**Comment by annezazu** [comment]
👋🏼 Is this issue still relevant? If yes, can we update it? If not, can we close out?

**Comment by jameskoster** [comment]
A lot of site building blocks still don't have previews. 

> So perhaps these block previews should just render the block itself (hopefully with the context passed to it). Is that an option?

This still feels like the path to explore.

**Comment by MaggieCabrera** [comment]
It seems like some of these are quick wins. I started with https://github.com/WordPress/gutenberg/pull/62937 and I didn't need to do much to get it to show. 

Some others like the prev/next are a little trickier, since it's one block that we want to have two different examples (it's the post-navigation-link block). I wonder if we could just show the title of the post instead for that one and keep one preview for both of them

**Comment by carolinan** [comment]
In January I started on a block that has these two links, next and prev as inner blocks, similar to how the query pagination block works, I suggest leaving this block and its variation until later and continue with the quick wins. https://github.com/WordPress/gutenberg/issues/55157

**Comment by MaggieCabrera** [comment]
Working on some of these I notice that the preview in the inserter and the one in Global Styles are a little different. It seems like on Global Styles we are adding some extra CSS to alter the appearance and it's not clear to me if they should both look the same or if the differences are ok. Also, some of these are looking broken (like the page list):

Page list

GS | Inserter
--- | ---
<img width="287" alt="Screenshot 2024-07-04 at 19 23 02" src="https://github.com/WordPress/gutenberg/assets/3593343/4d3f6fab-873b-4625-b3cc-7ad77e5af15a"> | <img width="649" alt="Screenshot 2024-07-04 at 19 24 09" src="https://github.com/WordPress/gutenberg/assets/3593343/dd9cd998-f6d8-4b0a-ac60-224fc0f46ba1">


Image

GS | Inserter
--- | ---
<img width="289" alt="Screenshot 2024-07-04 at 19 23 17" src="https://github.com/WordPress/gutenberg/assets/3593343/af31c360-39ff-49e3-8119-9e10bd4e70d4"> | <img width="674" alt="Screenshot 2024-07-04 at 19 23 56" src="https://github.com/WordPress/gutenberg/assets/3593343/520cfed7-3d9c-482f-a97b-15e09ca440e3">

Button

GS | Inserter
--- | ---
<img width="288" alt="Screenshot 2024-07-04 at 19 23 42" src="https://github.com/WordPress/gutenberg/assets/3593343/7982afe7-d9d8-42d0-ae42-548e83f74147"> | <img width="689" alt="Screenshot 2024-07-04 at 19 23 50" src="https://github.com/WordPress/gutenberg/assets/3593343/987a7220-1d7a-444d-ac0f-922ee12c7404">

**Comment by jameskoster** [comment]
My initial thought is they should be the same. What's the reason for different styling in global styles? Maybe @jasmussen knows?

**Comment by jasmussen** [comment]
They should be the same.

I believe that there may have previously been a difference in how the two are loaded; where one is iframed, and the other may not be (?).

**Comment by MaggieCabrera** [comment]
> They should be the same.
> 
> I believe that there may have previously been a difference in how the two are loaded; where one is iframed, and the other may not be (?).

yes, and also, the one in GS is a fixed height, whereas the inserter one has a flexible one, as you can clearly see on the page list example. I could open a PR to align the styles of the two while maintaining a scroll for the GS one.

Also, I'm thinking we want the blocks to be centered when possible, right?

**Comment by jasmussen** [comment]
> Also, I'm thinking we want the blocks to be centered when possible, right?

Yes, though that can be hard when it's larger than the preview. Not sure what flex magic might help there.

**Comment by MaggieCabrera** [comment]
For context, this is where we added the extra CSS that centers things in global styles: https://github.com/WordPress/gutenberg/pull/46727

**Comment by MaggieCabrera** [comment]
> > Also, I'm thinking we want the blocks to be centered when possible, right?
> 
> Yes, though that can be hard when it's larger than the preview. Not sure what flex magic might help there.

Seems like a CSS puzzle to me, bring it on :D

**Comment by mtias** [comment]
Closing this for now. If there are any blocks remaining we can tackle separately.
